[SDK50] Can't build Android app (:app:createBundleReleaseJsAndAssets FAILED)

### Minimal reproducible example

https://github.com/TomCorvus/ExpoMonorepo

### Summary

Hi,

Since SDK50, I can't build Android release build. I have an error:
```
> Task :app:createBundleReleaseJsAndAssets FAILED
CommandError: It looks like you're trying to use web support but don't have the required
dependencies installed.

Please install react-dom@18.2.0, @expo/metro-runtime@~3.1.1 by running:

npx expo install react-dom @expo/metro-runtime

If you're not using web, please ensure you remove the "web" string from the
platforms array in the project Expo config.


Execution failed for task ':app:createBundleReleaseJsAndAssets'.
> Process 'command 'node'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
BUILD FAILED in 44s
1057 actionable tasks: 22 executed, 1035 up-to-date
```

I have not the "web" string in Expo config file. This is a bare workflow project and app.json contains only app name.
I can build my app with debug keystore and debug mode but when I try to build it in release, it crashs.
So it's hard to give you a representative repro without giving you a release keystore.

I'm using `webpack` and `react-native-web` in my project only to generate the app storybook.
I don't know if it is the cause.

Thanks for your help.

### Environment

```
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.2.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.9.0 - /usr/local/bin/node
      Yarn: 4.0.2 - /usr/local/bin/yarn
      npm: 10.1.0 - /usr/local/bin/npm
      Watchman: 2023.07.10.00 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.14.3 - /Users/nickname/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.2, iOS 17.2, macOS 14.2, tvOS 17.2, visionOS 1.0, watchOS 10.2
      Android SDK:
        API Levels: 23, 26, 28, 29, 30, 31, 33, 34
        Build Tools: 28.0.3, 29.0.2, 29.0.3, 30.0.2, 30.0.3, 31.0.0, 33.0.0, 33.0.1, 34.0.0
        System Images: android-26 | Google APIs Intel x86 Atom, android-27 | Google APIs Intel x86 Atom, android-27 | Google Play Intel x86 Atom, android-28 | Google APIs Intel x86 Atom, android-29 | Google APIs Intel x86 Atom, android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom, android-31 | Google APIs ARM 64 v8a, android-31 | Google APIs Intel x86_64 Atom, android-33 | Google APIs Intel x86_64 Atom
    IDEs:
      Android Studio: 2023.1 AI-231.9392.1.2311.11255304
      Xcode: 15.2/15C500b - /usr/bin/xcodebuild
    npmPackages:
      expo: ^50.0.3 => 50.0.3 
      react: ^18.2.0 => 18.2.0 
      react-dom: ^18.2.0 => 18.2.0 
      react-native: ^0.73.2 => 0.73.2 
      react-native-web: ^0.19.10 => 0.19.10 
    Expo Workflow: bare
```

--- a/packages/@expo/cli/CHANGELOG.md
+++ b/packages/@expo/cli/CHANGELOG.md
@@ -10,7 +10,7 @@
 
 ### 🐛 Bug fixes
 
-- Add better validation for Metro web when exporting production bundles.
+- Add better validation for Metro web when exporting production bundles. ([#26732](https://github.com/expo/expo/pull/26732) by [@EvanBacon](https://github.com/EvanBacon))
 - Fix stack traces for Node.js errors. ([#26607](https://github.com/expo/expo/pull/26607) by [@EvanBacon](https://github.com/EvanBacon))
 - Fixed crash when launching React DevTools. ([#26550](https://github.com/expo/expo/pull/26550) by [@kudo](https://github.com/kudo))
 - Only show "Web is waiting" message after project is initialized with web. ([#26694](https://github.com/expo/expo/pull/26694) by [@byCedric](https://github.com/byCedric))

--- a/packages/@expo/cli/CHANGELOG.md
+++ b/packages/@expo/cli/CHANGELOG.md
@@ -10,6 +10,7 @@
 
 ### 🐛 Bug fixes
 
+- Add better validation for Metro web when exporting production bundles.
 - Fix stack traces for Node.js errors. ([#26607](https://github.com/expo/expo/pull/26607) by [@EvanBacon](https://github.com/EvanBacon))
 - Fixed crash when launching React DevTools. ([#26550](https://github.com/expo/expo/pull/26550) by [@kudo](https://github.com/kudo))
 - Only show "Web is waiting" message after project is initialized with web. ([#26694](https://github.com/expo/expo/pull/26694) by [@byCedric](https://github.com/byCedric))

--- a/packages/@expo/cli/src/export/__tests__/exportApp-test.ts
+++ b/packages/@expo/cli/src/export/__tests__/exportApp-test.ts
@@ -7,6 +7,10 @@ import { ExportAssetMap } from '../saveAssets';
 
 jest.mock('../../log');
 
+jest.mock('../../start/doctor/web/WebSupportProjectPrerequisite', () => ({
+  WebSupportProjectPrerequisite: jest.fn(() => ({ assertAsync: jest.fn() })),
+}));
+
 jest.mock('../exportStaticAsync', () => ({
   unstable_exportStaticAsync: jest.fn(
     async (_root: string, { files }: { files: ExportAssetMap }) => {

--- a/packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts
+++ b/packages/@expo/cli/src/export/__tests__/resolveOptions-test.ts
@@ -27,7 +27,7 @@ describe(resolveOptionsAsync, () => {
       exp: { web: { bundler: 'webpack' }, platforms: ['ios', 'android', 'web'] },
     });
     await expect(resolveOptionsAsync('/', { '--platform': ['web'] })).rejects.toThrow(
-      /^Platform "web" is not configured to use the Metro bundler in the project Expo config\./
+      /^Platform "web" is not configured to use the Metro bundler in the project Expo config,/
     );
   });
 

--- a/packages/@expo/cli/src/export/exportApp.ts
+++ b/packages/@expo/cli/src/export/exportApp.ts
@@ -13,6 +13,7 @@ import { Options } from './resolveOptions';
 import { ExportAssetMap, getFilesFromSerialAssets, persistMetroFilesAsync } from './saveAssets';
 import { createAssetMap, createSourceMapDebugHtml } from './writeContents';
 import * as Log from '../log';
+import { WebSupportProjectPrerequisite } from '../start/doctor/web/WebSupportProjectPrerequisite';
 import { getRouterDirectoryModuleIdWithManifest } from '../start/server/metro/router';
 import { serializeHtmlWithAssets } from '../start/server/metro/serializeHtml';
 import {
@@ -54,6 +55,10 @@ export async function exportAppAsync(
     // Web doesn't require validation.
     skipValidation: platforms.length === 1 && platforms[0] === 'web',
   });
+
+  if (platforms.includes('web')) {
+    await new WebSupportProjectPrerequisite(projectRoot).assertAsync();
+  }
 
   const useServerRendering = ['static', 'server'].includes(exp.web?.output ?? '');
   const baseUrl = getBaseUrlFromExpoConfig(exp);

--- a/packages/@expo/cli/src/export/resolveOptions.ts
+++ b/packages/@expo/cli/src/export/resolveOptions.ts
@@ -34,9 +34,15 @@ export function resolvePlatformOption(
 
   const assertPlatformBundler = (platform: Platform): Platform => {
     if (!platformsAvailable[platform]) {
+      if (!exp.platforms?.includes(platform) && platform === 'web') {
+        // Pass through so the more robust error message is shown.
+        return platform;
+      }
       throw new CommandError(
         'BAD_ARGS',
-        `Platform "${platform}" is not configured to use the Metro bundler in the project Expo config.`
+        `Platform "${platform}" is not configured to use the Metro bundler in the project Expo config, or is missing from the supported platforms in the platforms array: [${exp.platforms?.join(
+          ', '
+        )}].`
       );
     }
 

--- a/packages/@expo/cli/src/start/doctor/web/WebSupportProjectPrerequisite.ts
+++ b/packages/@expo/cli/src/start/doctor/web/WebSupportProjectPrerequisite.ts
@@ -98,7 +98,7 @@ export class WebSupportProjectPrerequisite extends ProjectPrerequisite {
 export function isWebPlatformExcluded(rootConfig: AppJSONConfig): boolean {
   // Detect if the 'web' string is purposefully missing from the platforms array.
   const isWebExcluded =
-    Array.isArray(rootConfig.expo?.platforms) &&
+    Array.isArray(rootConfig?.expo?.platforms) &&
     !!rootConfig.expo?.platforms.length &&
     !rootConfig.expo?.platforms.includes('web');
   return isWebExcluded;

--- a/packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts
+++ b/packages/@expo/cli/src/start/server/metro/withMetroMultiPlatform.ts
@@ -35,7 +35,6 @@ import { installExitHooks } from '../../../utils/exit';
 import { isInteractive } from '../../../utils/interactive';
 import { loadTsConfigPathsAsync, TsConfigPaths } from '../../../utils/tsconfig/loadTsConfigPaths';
 import { resolveWithTsConfigPaths } from '../../../utils/tsconfig/resolveWithTsConfigPaths';
-import { WebSupportProjectPrerequisite } from '../../doctor/web/WebSupportProjectPrerequisite';
 import { PlatformBundlers } from '../platformBundlers';
 
 type Mutable<T> = { -readonly [K in keyof T]: T[K] };
@@ -469,10 +468,6 @@ export async function withMetroMultiPlatformAsync(
   config.transformer._expoRouterWebRendering = webOutput;
   // @ts-expect-error: Invalidate the cache when the location of expo-router changes on-disk.
   config.transformer._expoRouterPath = resolveFrom.silent(projectRoot, 'expo-router');
-
-  if (exp.platforms?.includes('web') && platformBundlers.web === 'metro') {
-    await new WebSupportProjectPrerequisite(projectRoot).assertAsync();
-  }
 
   let tsconfig: null | TsConfigPaths = null;
 

